### PR TITLE
functional_collectives autograd support for all_to_all_single

### PR DESCRIPTION
**Pull request by d4l3k**
This adds autograd support for all_to_all_single. This allows using it in normal nn.Modules and backpropagating across machines. This should make it much easier to experiment with certain types of distributed operations such as Context Parallelism.

* Compilation works including with the AsyncCollectiveTensor wrapper forwards and backwards.
* I had to make some tweaks to the signature of the all_to_all_single native op as autograd Functions don't support List[int].
* I've verified forwards and backwards accuracy by using the DTensor ring flash attention implementation forwards with no modifications :partying_face: 

Test plan:

```sh
pytest test/distributed/test_functional_api.py -k test_all_to_all_single_compile
# added test that uses the ring_attention forwards implementation
pytest test/distributed/_tensor/test_attention.py -k autograd_is_causal
```

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire

### CODE REVIEW

File: torch/distributed/_functional_collectives.py [hunk 1]
```diff
@@ -625,8 +660,10 @@ def wrap(e: torch.Tensor):
         unwrapped_args = tree_map_only(AsyncCollectiveTensor, unwrap, args)
         unwrapped_kwargs = tree_map_only(AsyncCollectiveTensor, unwrap, kwargs)
 
-        # we don't wrap the result as it doesn't need to be waited on.
-        out = func(*unwrapped_args, **unwrapped_kwargs)
+        # we need this so that the grad_fn is correctly propagated
+        with torch.overrides.enable_reentrant_dispatch():
```
Comment (wconstab) [inline torch/distributed/_functional_collectives.py:664]: what does this do?

Comment (d4l3k) [reply torch/distributed/_functional_collectives.py:664]: I don't really understand what it's doing under the hood but it makes the grad_fn be set correctly

Found it from this example with a tensor subclass + autograd https://github.com/pytorch/pytorch/blob/c8e117fb761235296a7eae6f88bce4ab68ab5c2a/torch/testing/_internal/logging_tensor.py#L94

It seems to do something with thread local storage but it's unclear why that is needed here https://github.com/pytorch/pytorch/blob/c8e117fb761235296a7eae6f88bce4ab68ab5c2a/aten/src/ATen/core/PythonFallbackKernel.h#L7

cc @albanD

File: torch/distributed/_functional_collectives.py [hunk 2]
```diff
@@ -848,15 +885,18 @@ def mk_out_tensor(input):
 # but then you pass those sizes explicitly, and the all to all itself
 # isn't dynamic, it just follows the specified output splits
 def _all_to_all_single_meta(
-    input, output_split_sizes, input_split_sizes, *args, **kwargs
+    output_size: int,
+    input,
+    output_split_sizes: torch.Tensor,
+    input_split_sizes: torch.Tensor,
+    *args,
+    **kwargs,
```
Comment (yifuwang) [inline torch/distributed/_functional_collectives.py:893]: Chatted offline. The modification may not be necessary if the autograd function is called within the dispatcher op.

**Comment by d4l3k** [comment]
Switched to C++ implementation at https://github.com/pytorch/pytorch/pull/123599
